hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -268,20 +276,36 @@ protected List<DatanodeDetails> chooseDatanodesInternal(
     int numberOfRacksRequired = getRequiredRackCount(requiredReplicationFactor);
     int additionalRacksRequired =
```

Comment:
From what I understand all we have to change is make additionalRacksRequired = Math.min(nodesRequired, numberOfRacksRequired - usedRacksCntMap.size()); and remove the condition check nodesRequired < additionalRacksRequired all together. We wouldn't need a fall back flag then. This particular issue can only happen when under replication handler is calling this function asking for lesser nodes than the misreplication count. Rackscatter Placement policy is ensuring that the misreplication count is going down if an additional node is chosen.

Reply:
On the misreplication handler in that case we should only copy the number of replicas by the change in the misreplication count. Does this make sense?

Reply:
> From what I understand all we have to change is make additionalRacksRequired = Math.min(nodesRequired, numberOfRacksRequired - usedRacksCntMap.size()); and remove the condition check nodesRequired < additionalRacksRequired all together.

That will work for this particular case, yes. However simply doing that will not work in the first case, which is:
```
  /**
   * Scenario:
   * rack0 -> node0
   * rack1 -> node1
   * rack2 -> node2
   * rack3 -> node3
   * rack4 -> node4, node5
   * <p>
   * node0, node1, node2, node4 are used nodes. node3 is excluded node.
   * Expectation is that node5 should be chosen.
```

So, we'll still need to "fall back and relax the policy" somehow, which is why I decided to use the existing `fallback` flag.

Reply:
yeah this does makes sense when the rack itself is not available. Thanks for pointing out such a scenario. @siddhantsangwan

Reply:
I had another question. Ideally this can happen only in a small cluster. Shouldn't we ideally wait for the unhealthy replica to get deleted and create the replica ? That should be fine right?

Reply:
> Ideally this can happen only in a small cluster.

Yes, all of these are edge cases in small cluster situations.

> Shouldn't we ideally wait for the unhealthy replica to get deleted and create the replica ? That should be fine right?

Yes, that's the idea that's already existing in `ECUnderReplicationHandler`. This change as it stands should not interfere with that.